dali/pipeline/operators/crop/basic/coords.h

```diff
@@ -0,0 +1,109 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_OPERATORS_CROP_BASIC_COORDS_H_
+#define DALI_PIPELINE_OPERATORS_CROP_BASIC_COORDS_H_
+
+#include <array>
+#include <cstdint>
+#include <iostream>
+
+namespace dali {
+namespace basic_crop {
+
+namespace detail {
+
+template <typename T, size_t N>
+int64_t getOffsetImplLinear(const T& shape, std::array<int64_t, N> coords) {
+  int64_t result = coords[0];
+  for (size_t i = 1; i < N; i++) {
+    result *= shape[i];
+    result += coords[i];
+  }
+  return result;
+}
+
+}  // namespace detail
+
+template <int64_t... Ints>
+class dali_index_sequence {};
+
+template <typename T>
+struct index_to_array {};
+
+template <int64_t... Ints>
+struct index_to_array<dali_index_sequence<Ints...>> {
+  static constexpr std::array<int64_t, sizeof...(Ints)> array = {Ints...};
+  static constexpr size_t N = sizeof...(Ints);
+};
+
+/**
+ * @brief Permute shape by given order
+ *
+ * @tparam order
+ * @param shape
+ * @return std::array<int64_t, sizeof...(order)>
+ */
+template <int64_t... order>
+std::array<int64_t, sizeof...(order)> permuteShape(
+    const std::array<int64_t, sizeof...(order)>& shape, dali_index_sequence<order...> = {}) {
+  return {shape[order]...};
+}
+
+/**
+ * @brief Get the offset to coordiantes after requested permutation
+ *
+ * For given shape {s_1, s_2, ... s_n} and coordinates (c_1, c_2, ..., c_n)
+ * return offset after doing axes permuation <o_1, o_2, ..., o_n> (at position i we place axis o_i).
+ *
+ * calculates:
+ *   c[o_1] * (s[2] * s[3] * ... * s[n])
+ * + c[o_2] * (s[3] * s[4] * ... * s[n])
+ * + ...
+ * + c[o_{n-1}] * (s[n])
+ * + c[o_n]
+ *
+ * TODO(klecki) formal explanation and handle default case
+ *
+ * Example: for given shape {N, H, W, C}, and coordiantes (n, h, w, c),
+ * and order <0, 1, 2, 3> it would calculate offsets for NHWC data layout.
+ * Other order will produce offsets to permutation of dimensions, for example:
+ * <0, 3, 1, 2> will result in offsets to NCHW data layout.
+ *
+ * User may provide an optional argument of dali_index_sequence object that carries the information
+ * of the order instead of specifying it explicitly in template arguments.
+ *
+ * getOffset<0, 3, 1, 2>({4, 1080, 1920, 3}, 0, 500, 100, 2);
+ * is equivalent to:
+ * getOffset({4, 1080, 1920, 3}, 0, 500, 100, 2, dali_index_sequence<0, 3, 1, 2>{});
+ *
+ * @tparam order Permutation of input dimensions
+ * @tparam T type of container carrying shape information
+ * @param shape Data layout "before" permutation
+ * @param coords Coordinates "after" permutation
+ * @return int64_t offset to given coordinate
+ */
+template <int64_t... order, typename T>
+int64_t getOffset(const T& shape, std::array<int64_t, sizeof...(order)> coords,
```

Comment:
Please provide unit tests for this function

Reply:
I removed it, as this was dead code.

Reply:
Done.